Display a route ladder for each selected route

arkadyan opened the pull request:
Asana ticket: [build UI for route ladder - name in center (wireframe-level version)](https://app.asana.com/0/1112935048846093/1114820119722043)

![Screen Shot 2019-04-11 at 13 25 19](https://user-images.githubusercontent.com/42339/55978450-d9f5f680-5c5d-11e9-8af4-6939cdebde49.png)

skyqrose commented on assets/src/skate.d.ts:
```diff
@@ -3,3 +3,13 @@ export interface Route {
+export type TimepointId = string
+
+export interface TimepointsByRoute {
+  [routeId: string]: Timepoint[]
```
Can be the `RouteId` type instead of `string`.

arkadyan replied:
Typescript doesn't allow index signature parameter types to be type aliases.

skyqrose commented on assets/css/app.scss:
```diff
@@ -1 +1,50 @@
+  display: flex;
+}
+
+.m-route-ladders {
```
Let's put this in a `route_ladders.scss` file

skyqrose commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,59 @@
+interface Props {
+  route: Route
+  dispatch: Dispatch
+  timepointsByRoute: TimepointsByRoute
```
```suggestion
  timepoints: Timepoint[] | null
```
If we do the lookup before we call this component, then it doesn't need to know anything about other ladders.

arkadyan replied:
👍 I actually started to do this at one point but it got lost in the shuffle. Added a function.

skyqrose commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,59 @@
+        dispatch(setTimepointsForRoute(route.id, newTimepoints))
+      )
+    }
+  })
```
This effect needs something to make sure it's not called multiple times, since it could be rendered multiple times before we get the response.

Alternatively, maybe this should live in the reducer, in the case for `selectRoute`. That's how Elm does it: the reducer both updates the state and initiates any new side effects. I don't know if the React convention is to put side effects in the components or the reducer.

arkadyan replied:
Good call. I added a null state for loading and updated the check to only care about `undefined`.

Hooks need to be in React components so we can't move it to the reducer, but we should keep in mind the ability to create custom hooks if we need to share or abstract them in the future. (I don't think there's value in doing that here at this point.)

skyqrose commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,59 @@
+}
+
+const TimepointStop = () => (
+  <div className="m-route-ladder__timepoint-stop">
```
```suggestion
  <div className="m-route-ladder__stop">
```
BEM doesn't need the whole hierarchy. I don't think having `timepoint-` adds any understanding, so I'd lean slightly against including it.

(`__timepoint-name` does help disambiguate it from `__route-name`, so this comment wouldn't apply there.)

skyqrose commented on assets/tests/components/routeLadders.test.tsx:
```diff
@@ -0,0 +1,51 @@
+      { id: "DUDLY" },
+      { id: "LOUIS" },
+      { id: "MALCX" },
+      { id: "RUGG" },
```
We don't need this much sample data. 2 or 3 for each route would be fine. Using less data would make it easier to add fields to the timepoint type, and easier to scroll through this file once it has lots of tests.

(Same for `routeLadder.test.tsx`)

arkadyan commented on assets/src/skate.d.ts:
```diff
@@ -3,3 +3,13 @@ export interface Route {
+export type TimepointId = string
+
+export interface TimepointsByRoute {
+  [routeId: string]: Timepoint[]
```
Re `routeId`: typescript doesn't allow index signature parameter types to be type aliases.

Re `TimepointsByRouteId`: 👍

arkadyan commented on assets/css/app.scss:
```diff
@@ -1 +1,50 @@
+
+.m-route-ladder__timepoint-stop {
+  height: 60px;
+  width: 20px;
```
I was trying to match the svg attributes which are in `px`, but I discovered that if I set the `line-height` instead then I don't need to set `height` and `width`.

arkadyan commented:
@skyqrose I added a test to verify that we fetch timepoint data when we render a route ladder for which we don't already have data. I think I've addressed the rest of the feedback as well.

skyqrose commented on assets/tests/components/routeLadder.test.tsx:
```diff
@@ -0,0 +1,114 @@
+declare global {
+  interface Window {
+    /* eslint-disable typescript/no-explicit-any */
+    fetch: (uri: string) => Promise<any>
```
Could this be `Promise<Response>` in order to avoid the `eslint-disable`? Or would that require specifying fields in the mock that we don't care about?

arkadyan replied:
It appears that you have to add a bunch of extra functions we don't care about to support that, so I think it's not worth it…

skyqrose commented on assets/src/state.ts:
```diff
@@ -66,10 +109,35 @@ export const reducer = (state: State, action: Action): State => {
+
+export const timepointsForRouteId = (
+  state: State
+): TimepointsForRouteId => routeId => state.timepointsByRoute[routeId]
```
What's the difference between the names `timepointsByRoute`, `TimepointsByRouteId` and `[tT]imepointsForRouteId`? Could they all have the same name?

Is using a function to wrap the object so that we can get a real type on the function because of the weird typing restriction on objects?

arkadyan replied:
Fixed all this as discussed.

skyqrose replied:
Just so it's written down:
"timepoints for route" is for functions / actions etc that work on a single route.
"timepoints by route id" is the map of many route ids to their timepoints.
"timepoints for route id" no longer exists.
"timepoints by route" no longer exists.

skyqrose commented on assets/tests/components/routeLadder.test.tsx:
```diff
@@ -0,0 +1,114 @@
+  window.fetch = () =>
+    Promise.resolve({
+      json: () => ({
+        data: [{ id: "28" }, { id: "39" }, { id: "71" }],
```
This looks like `Route[]`, not `Timepoint[]`. The endpoint we're calling should be a list of timepoints.
(Same for the other test.)

arkadyan replied:
Oops, yeah…

skyqrose commented on assets/tests/components/routeLadder.test.tsx:
```diff
@@ -0,0 +1,114 @@
+      .toJSON()
+  })
+
+  expect(mockDispatch.mock.calls.length).toBe(1)
```
Shouldn't we have two calls? One to say we're loading, and a second one once the api call returns?

arkadyan replied:
The second one happens asynchronously so the test finishes before that happens. But I tested the reducer to test handling that data.

skyqrose replied:
If the promise doesn't resolve, before the test finishes, does its return value matter?

Is there a way to make the test wait for the promise? This page has some options: https://jestjs.io/docs/en/asynchronous

arkadyan replied:
Trouble is we aren't calling the promise, we are triggering it by interacting with the DOM. But I think demostrating that the fetch process starts seems like a sufficient test here to me.

arkadyan commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,66 @@
+interface Props {
+  route: Route
+  dispatch: Dispatch
+  timepointsForRouteId: TimepointsForRouteId
```
No this is a function to retrieve the timepoints given a route id, not the full object.

arkadyan replied:
Replaced this as we discussed.

arkadyan commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,66 @@
+        dispatch(setTimepointsForRoute(route.id, newTimepoints))
+      )
+    }
+  }, [])
```
This doesn't prevent duplicate calls, the `timepoints === undefined` check above does.

skyqrose commented on assets/tests/components/routeLadders.test.tsx:
```diff
@@ -0,0 +1,26 @@
+  const timepointsByRouteId: TimepointsByRouteId = {
+    "1": [{ id: "WASMA" }, { id: "MELWA" }, { id: "HHGAT" }],
+    "28": [{ id: "MATPN" }, { id: "WELLH" }, { id: "MORTN" }],
+  }
```
Is it worth testing a `null` case and an `undefined` case here?

skyqrose commented on assets/tests/components/routeLadder.test.tsx:
```diff
@@ -0,0 +1,101 @@
+      }),
+      ok: true,
+      status: 200,
+    })
```
This test shouldn't call fetch, so this isn't needed.

Alternatively, for this test and the previous, could we make `window.fetch` a `jest.fn` so we can check that it was called the right number of times?

skyqrose commented on assets/src/components/routeLadder.tsx:
```diff
@@ -0,0 +1,63 @@
+
+interface Props {
+  route: Route
+  timepoints: LoadableTimepoints | undefined
```
Thoughts on moving this `|` to `skate.d` so that the `undefined`=not asked and `null`=loading concepts are documented in one place?

One way to do that might be to change `LoadableTimepoints` to `Timepoint[] | null | undefined`, and then `TimepointsByRouteId` use `Timepoint[] | null` instead of `LoadableTimepoints`

paulswartz commented on assets/src/skate.d.ts:
```diff
@@ -8,9 +8,9 @@ export interface Timepoint {
-export type LoadableTimepoints = Timepoint[] | null
+// An undefined value indicates that the timepoints need to be loaded
+// A null value indicates that we are currently loading the timepoints
+export type LoadableTimepoints = Timepoint[] | null | undefined
```
Future change: have an actual type for these cases rather than overriding null and undefined.

arkadyan replied:
Yeah, I think that could be worthwhile.
